doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,894 @@
+
+This guide walks through the setup necessary to run mypy, a static type checker, on client library code. It also
+contains some general typing tips and guidance as it relates to common types/patterns we use in the Python SDK.
+
+### Table of contents
+
+- [Intro to typing in Python](TODO)
+- [Typing a client library](TODO)
+- [Types usable in annotations](TODO)
+- [Install and run mypy on your client library code](TODO)
+- [Typing tips and guidance for the Python SDK](TODO)
+    - [Debug mypy with reveal_type() and reveal_locals()](TODO)
+    - [Use typing.Any sparingly](TODO)
+    - [Use typing.Union when accepting more than one type](TODO)
+    - [Use typing.Optional when a parameter default is None](TODO)
+    - [Typing for collections](TODO)
+        - [Mapping types](TODO)
+        - [List](TODO)
+        - [Tuple](TODO)
+    - [Passing a function or a class as a parameter or return type](TODO)
+    - [Typing variadic arguments - *args and **kwargs](TODO)
+    - [Use forward references when the type is not defined yet](TODO)
+    - [Use typing.overload to overload a function](TODO)
+    - [Use typing.cast to help mypy understand a type](TODO)
+    - [Use typing.Protocol for structural subtyping](TODO)
+    - [Use TypeVar for generic type hinting](TODO)
+    - [Use AnyStr when your parameter accepts both str and bytes](TODO)
+    - [Use typing.TypeAlias when creating a type alias](TODO)
+    - [Use TYPE_CHECKING to avoid circular imports](TODO)
+- [How to ignore a mypy typing error](TODO)
+- [How to opt out of mypy type checking](TODO)
+
+## Intro to typing in Python
+
+Python is a dynamically typed (or sometimes called "duck typed") language. This means that the interpreter only checks
+types as code runs and types are allowed to change. This is different from more statically typed languages, like Java,
+where you will have you declare types in code and check them at compile time. However,
+with [PEP 484](https://peps.python.org/pep-0484/), type hints were introduced to Python which makes it possible to do
+the static type checking of Python code. Type hints can be written into Python code to indicate the type of a variable,
+argument, return type, etc. There are tools available to perform static type checking using type hints,
+like [mypy](https://mypy.readthedocs.io/en/stable/). Note that type hints have no effect on the code at runtime and are
+not enforced by the interpreter.
+
+There are some key benefits to using and checking type hints in Python code:
+
+1) type hints checked by a static type checker can help developers find bugs in Python code (type hints can be thought
+   of as "free" unit tests, although they do not replace a thorough test suite).
+2) type hints are shipped in our packages so that customers can consume them to further type check their own
+   applications and get improved IDE and Intellisense experiences
+
+There are two styles of type hints: comment and annotation (preferred).
```

Comment:
Don't do comment anymore. It's not prefered, it's comments are dead with Python 2